### PR TITLE
[SPARK-17693] [SQL] Fixed Insert Failure To Data Source Tables when the Schema has the Comment Field

### PR DESCRIPTION
**Pull request by gatorsmile**
### What changes were proposed in this pull request?

``` SQL
CREATE TABLE tab1(col1 int COMMENT 'a', col2 int) USING parquet
INSERT INTO TABLE tab1 SELECT 1, 2
```

The insert attempt will fail if the target table has a column with comments. The error is strange to the external users:

```
assertion failed: No plan for InsertIntoTable Relation[col1#15,col2#16] parquet, false, false
+- Project [1 AS col1#19, 2 AS col2#20]
   +- OneRowRelation$
```

This PR is to fix the above bug by ignoring the metadata when comparing the schema between the table and the query.
### How was this patch tested?

Added a test case

### CODE REVIEW

**Comment by gatorsmile** [comment]
cc @cloud-fan @yhuai

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala [hunk 1]
```diff
@@ -159,8 +159,9 @@ case class DataSourceAnalysis(conf: CatalystConf) extends Rule[LogicalPlan] {
 
 
     case i @ logical.InsertIntoTable(
-           l @ LogicalRelation(t: HadoopFsRelation, _, _), part, query, overwrite, false)
-        if query.resolved && t.schema.asNullable == query.schema.asNullable =>
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala]: That's why I left a TODO and ask if we do need to rename: https://github.com/apache/spark/pull/15266/files#diff-73bd90660f41c12a87ee9fe8d35d856aL246

INSERT INTO TABLE is resolved by ordinal, so the field names don't matter. How about we don't rename and check `t.schema.map(_.dataType.asNullable) == query.map(_.dataType.asNullable)` here?

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala]: I tried it. CSV works, but Parquet and Json do not work. The input schema needs to be consistent with the existing table schema. That is why we still need to rename the column names.

Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala]: ah i see. But we only rename the top level fields, what if the inner struct field names don't match?

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala]: uh... A good catch! Let me try the complex types.

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala]: Did an investigation. For the struct type, we store the nested schema in the data type. Thus, after we cast the data type of struct, we also copy the schema at the same time. 

Below is the explained plan of the example I just added.
```
== Analyzed Logical Plan ==
InsertIntoHadoopFsRelationCommand file:/Users/xiaoli/IdeaProjects/sparkDelivery/spark-warehouse/tab2, ParquetFormat, <function0>, Map(path -> file:/Users/xiaoli/IdeaProjects/sparkDelivery/spark-warehouse/tab2), Append
   +- Project [cast(s#36 as struct<c:string,d:string>) AS p#40]
      +- Project [s#36]
         +- SubqueryAlias tab1
            +- Relation[s#36] parquet
```

**Comment by gatorsmile** [comment]
retest this please

**Comment by gatorsmile** [comment]
Any other comment? @cloud-fan Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala [hunk 2]
```diff
@@ -252,6 +251,8 @@ case class PreprocessTableInsertion(conf: SQLConf) extends Rule[LogicalPlan] {
         if (expected.dataType.sameType(actual.dataType) && expected.name == actual.name) {
           actual
         } else {
+          // Renaming is needed for handling the INSERT statement like
+          // INSERT INTO TABLE tab1 SELECT 1, 2
           Alias(Cast(actual, expected.dataType), expected.name)()
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala]: What if we copy the metadata during alias? Does it work too?

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala]: Submitted another PR: https://github.com/apache/spark/pull/15615. Could you check which one is better? Thank you!

**Comment by gatorsmile** [comment]
Since we prefer another solution in https://github.com/apache/spark/pull/15615, close this one now. Thanks!
